File: Leetcode/stack_queue/402_RemoveKDigits/Solution.java
class Solution {
    public String removeKdigits(String num, int k) {
        Deque<Character> dq = new ArrayDeque<>();
        for(int i = 0; i<num.length(); ++i){
            char temp = num.charAt(i);
            while(!dq.isEmpty() && dq.peekLast()>temp && k>0){
                dq.pollLast();
                k--;
            }
            dq.add(temp);
        }
        while(!dq.isEmpty() && k>0){
            dq.pollLast(); k--;
        }
        while(!dq.isEmpty() && dq.peekFirst()=='0'){
            dq.pollFirst();
        }
        if(dq.isEmpty()) return "0";
        String res = "";
        while(!dq.isEmpty()) res += dq.pollFirst();
        return res;
    }
}

File: Leetcode/stack_queue/402_RemoveKDigits/Solution.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        stack = []
        for i in range(0, len(num)):
            ch = num[i]
            while stack and int(stack[-1]) > int(ch) and k>0:
                stack.pop()
                k -= 1
            stack.append(ch)
        while stack and k>0:
            stack.pop()
            k-=1
        while stack and stack[0]=='0':
            stack.pop(0)
        res = "".join(stack)
        return "0" if res=="" else res
